specs/makeStyles.md

```diff
@@ -0,0 +1,290 @@
+
+::Risk:: What if we will not be able to describe the theme changes using just different values?
+::Risk:: A theme consist of hundreds or even thousands of tokens. We need to verify browsers can handle this amount of CSS variables.
+
+#### IE11 compatibility
+
+IE11 does not support CSS variables. If we decide to support IE11, we will split the styles to the items which use tokens and those which do not. The ones which do not use tokens will be processed build time, for the ones with tokens, we will fall back to runtime evaluation.
+
+### Hash-based classnames
+
+We will use a hash (MurmurHash2) of the CSS property name and the value to generate a classname. That results in deterministic (=stable) classnames. With that we can support server side rendering.
+
+### Passing overrides to children using `className`
+
+To avoid the ambivalence of using `className` vs `styles` props when passing overrides to children based on their type, we will use the dictionary sideeffect described in `useCSS` approach above.
+Parent component passes its overrides to a child as a list of classes in `className` prop. The child, takes the input classes and references the global dictionary to get back the necessary CSS attribute names to be able to perform the merge.
+
+#### Alternative 1: encode the CSS property name to the hash
+
+Instead of using the global dictionary side effect, we can use part of the hash in the classname to encode the CSS property:
+![](makeStyles-classname.png)
+
+Then when merging in a child, we can merge the classes by replacing previous classnames with the same CSS property hash.
+This approach makes the classnames longer and according to the test does not make the merging any faster.
+
+#### Alternative 2: use special `styles` prop for overrides
+
+If the global dictionary does not perform as expected, we might return back to the previous v0 approach where we use a dedicated `styles` prop to pass overrides from parent to child.
+
+### Another libraries using similar approach
+
+This has been inspired by [Facebook stylex talk](https://www.youtube.com/watch?v=9JZHodNR184).
+Similar approach is also used in Attlasian `compiled` v5.
+
+## Sample Code
+
+```jsx
+
+// Button.tsx
+const useButtonStyles = makeStyles([
+	[null, { cursor: 'pointer' }], // always applied
+  [ props => props.primary, tokens => ({ backgroundColor: tokens.colors.primary.background })],
+  [ props => pros.disabled, { backgroundColor: 'gray' }]
+]);
+
+const Button = ({className, ...props}) => {
+	const classes = useButtonStyles(props, className);
+	return <button className={classes} {...props} />;
+}
+
+// ButtonUsage.tsx
+const useButtonOverrides = makeStyles([
+  [null, { backgroundColor: 'red' ],
+]);
+
+const ButtonUsage = () => {
+  const buttonOverrides = useButtonOverrides({});
+
+  return <Button classname={buttonOverrides}>I have always red background</Button>
+}
+```
+
```

Comment:
Should we modify an example to show `ax()` function usage?

https://github.com/microsoft/fluentui/pull/16411